{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating OncoPrint Data Files\n",
    "\n",
    "The script will process all variant files and output files in an ingestible format for the R OncoPrint function.\n",
    "\n",
    "It will output oncoprint data for both replicate files and the merged variant callsets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>phyloP20way_mammalian</th>\n",
       "      <th>phastCons7way_vertebrate</th>\n",
       "      <th>phastCons20way_mammalian</th>\n",
       "      <th>SiPhy_29way_logOdds</th>\n",
       "      <th>Otherinfo</th>\n",
       "      <th>final_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>het</th>\n",
       "      <th>quality</th>\n",
       "      <th>depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>874779</td>\n",
       "      <td>874826</td>\n",
       "      <td>CCTCCCCAGCCACGGTGAGGACCCACCCTGGCATGATCCCCCTCATCA</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonframeshift deletion</td>\n",
       "      <td>SAMD11:NM_152486:exon7:c.645_692del:p.G220Dfs*447</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>het\\t.\\t20</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>26510311</td>\n",
       "      <td>26510311</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>exonic</td>\n",
       "      <td>CNKSR1</td>\n",
       "      <td>.</td>\n",
       "      <td>frameshift deletion</td>\n",
       "      <td>CNKSR1:NM_001297647:exon9:c.866delC:p.P291Hfs*...</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>het\\t.\\t31</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>158576883</td>\n",
       "      <td>158576883</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>OR10Z1</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>OR10Z1:NM_001004478:exon1:c.G655A:p.A219T</td>\n",
       "      <td>...</td>\n",
       "      <td>0.949</td>\n",
       "      <td>0.822</td>\n",
       "      <td>0.761</td>\n",
       "      <td>4.671</td>\n",
       "      <td>het\\t.\\t127</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>127</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr      Start        End                                               Ref  \\\n",
       "0   1     874779     874826  CCTCCCCAGCCACGGTGAGGACCCACCCTGGCATGATCCCCCTCATCA   \n",
       "1   1   26510311   26510311                                                 C   \n",
       "2   1  158576883  158576883                                                 G   \n",
       "\n",
       "  Alt Func.refGene Gene.refGene GeneDetail.refGene      ExonicFunc.refGene  \\\n",
       "0   -       exonic       SAMD11                  .  nonframeshift deletion   \n",
       "1   -       exonic       CNKSR1                  .     frameshift deletion   \n",
       "2   A       exonic       OR10Z1                  .       nonsynonymous SNV   \n",
       "\n",
       "                                    AAChange.refGene  ...   \\\n",
       "0  SAMD11:NM_152486:exon7:c.645_692del:p.G220Dfs*447  ...    \n",
       "1  CNKSR1:NM_001297647:exon9:c.866delC:p.P291Hfs*...  ...    \n",
       "2          OR10Z1:NM_001004478:exon1:c.G655A:p.A219T  ...    \n",
       "\n",
       "  phyloP20way_mammalian phastCons7way_vertebrate phastCons20way_mammalian  \\\n",
       "0                     .                        .                        .   \n",
       "1                     .                        .                        .   \n",
       "2                 0.949                    0.822                    0.761   \n",
       "\n",
       "  SiPhy_29way_logOdds    Otherinfo   final_id sample_name  het quality depth  \n",
       "0                   .   het\\t.\\t20  008-M2-F0      019-F0  het       .    20  \n",
       "1                   .   het\\t.\\t31  008-M2-F0      019-F0  het       .    31  \n",
       "2               4.671  het\\t.\\t127  008-M2-F0      019-F0  het       .   127  \n",
       "\n",
       "[3 rows x 60 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load all cosmic variants called in this dataset\n",
    "# This file was generated in filter_variants.ipynb\n",
    "cosmic_all_file = os.path.join('results', 'all_cosmic_variants.tsv')\n",
    "cosmic_all_df = pd.read_table(cosmic_all_file)\n",
    "cosmic_all_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NCOR2                39\n",
       "OR11G2               29\n",
       "KRAS                 29\n",
       "MUC4;MUC4            28\n",
       "DNHD1                27\n",
       "TSPYL1               26\n",
       "CBWD1                25\n",
       "GOLGA8M              24\n",
       "MARF1                24\n",
       "ARHGEF11;ETV3L       24\n",
       "WDR66                24\n",
       "PHF2                 24\n",
       "LCE4A                23\n",
       "GOLGA6D              22\n",
       "KRTAP5-11;FAM86C1    21\n",
       "CGB8                 21\n",
       "ACIN1                21\n",
       "POTEI                21\n",
       "MROH8;MROH8          20\n",
       "BRCA1                20\n",
       "Name: Gene.refGene, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the 50 most commonly altered COSMIC genes?\n",
    "top_n = 50\n",
    "paad_genes = cosmic_all_df['Gene.refGene'].value_counts().head(top_n).index.tolist()\n",
    "cosmic_all_df['Gene.refGene'].value_counts().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate OncoPrint Data\n",
    "\n",
    "### For All Replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 15.7 s, sys: 472 ms, total: 16.2 s\n",
      "Wall time: 16.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# Process each replicate by observed COSMIC mutation# Proce \n",
    "variant_file_path = os.path.join('results', 'processed_vcfs')\n",
    "variant_assign = []\n",
    "case_id = []\n",
    "for variant_file in os.listdir(variant_file_path):\n",
    "    # Load and subset file to only variants in the COSMIC db\n",
    "    variant_df = pd.read_table(os.path.join(variant_file_path, variant_file), index_col=0)\n",
    "    variant_sub_df = variant_df[variant_df['cosmic70'] != '.']\n",
    "    \n",
    "    # Define mutated genes if they exist for the given variant\n",
    "    variant_class = ['MUT;' if x in variant_sub_df['Gene.refGene'].tolist() else ''\n",
    "                     for x in paad_genes]\n",
    "    \n",
    "    # Store results\n",
    "    variant_assign.append(variant_class)\n",
    "    case_id.append(variant_file.replace('_001_processed_variants.tsv.bz2', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate and save oncoprint data for all replicates\n",
    "oncoprint_file = os.path.join('results', 'oncoprint_replicates.tsv')\n",
    "oncoprint_df = pd.DataFrame(variant_assign, index=case_id, columns=paad_genes).sort_index()\n",
    "oncoprint_df.index.name = 'Case.ID'\n",
    "oncoprint_df.to_csv(oncoprint_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCOR2</th>\n",
       "      <th>OR11G2</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>MUC4;MUC4</th>\n",
       "      <th>DNHD1</th>\n",
       "      <th>TSPYL1</th>\n",
       "      <th>CBWD1</th>\n",
       "      <th>GOLGA8M</th>\n",
       "      <th>MARF1</th>\n",
       "      <th>ARHGEF11;ETV3L</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF676</th>\n",
       "      <th>DPCR1</th>\n",
       "      <th>TP53</th>\n",
       "      <th>FAM161B</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>SMAD4</th>\n",
       "      <th>AGFG1</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PRELID3B</th>\n",
       "      <th>ADRA2B;ADRA2B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case.ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>001-F0_S3_L001</th>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>001-F0_S3_L002</th>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>001-F0_S3_L003</th>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               NCOR2 OR11G2  KRAS MUC4;MUC4 DNHD1 TSPYL1 CBWD1 GOLGA8M MARF1  \\\n",
       "Case.ID                                                                        \n",
       "001-F0_S3_L001  MUT;   MUT;  MUT;      MUT;  MUT;   MUT;  MUT;          MUT;   \n",
       "001-F0_S3_L002         MUT;  MUT;            MUT;   MUT;  MUT;          MUT;   \n",
       "001-F0_S3_L003  MUT;   MUT;  MUT;      MUT;  MUT;   MUT;  MUT;          MUT;   \n",
       "\n",
       "               ARHGEF11;ETV3L      ...      ZNF676 DPCR1 TP53 FAM161B GNAS  \\\n",
       "Case.ID                            ...                                       \n",
       "001-F0_S3_L001           MUT;      ...                           MUT;        \n",
       "001-F0_S3_L002           MUT;      ...              MUT;         MUT;        \n",
       "001-F0_S3_L003           MUT;      ...              MUT;         MUT;        \n",
       "\n",
       "               SMAD4 AGFG1 PIK3CA PRELID3B ADRA2B;ADRA2B  \n",
       "Case.ID                                                   \n",
       "001-F0_S3_L001                        MUT;                \n",
       "001-F0_S3_L002  MUT;                                      \n",
       "001-F0_S3_L003  MUT;                  MUT;                \n",
       "\n",
       "[3 rows x 50 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oncoprint_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merged Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_assign_consensus = []\n",
    "case_id_consensus = []\n",
    "for sample_id in set(cosmic_all_df['final_id']):\n",
    "    # Subset file to given sample ID\n",
    "    variant_sub_df = cosmic_all_df.query('final_id == @sample_id')\n",
    "    \n",
    "    # Define mutated genes if they exist for the given variant\n",
    "    variant_class = ['MUT;' if x in variant_sub_df['Gene.refGene'].tolist() else ''\n",
    "                     for x in paad_genes]\n",
    "    \n",
    "    # Store results\n",
    "    variant_assign_consensus.append(variant_class)\n",
    "    case_id_consensus.append(sample_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate and save oncoprint data for consensus samples\n",
    "oncoprint_consensus_file = os.path.join('results', 'oncoprint_merged.tsv')\n",
    "\n",
    "oncoprint_consensus_df = (\n",
    "    pd.DataFrame(variant_assign_consensus,\n",
    "                 index=case_id_consensus,\n",
    "                 columns=paad_genes)\n",
    "    )\n",
    "oncoprint_consensus_df.index.name = 'Case.ID'\n",
    "oncoprint_consensus_df.to_csv(oncoprint_consensus_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCOR2</th>\n",
       "      <th>OR11G2</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>MUC4;MUC4</th>\n",
       "      <th>DNHD1</th>\n",
       "      <th>TSPYL1</th>\n",
       "      <th>CBWD1</th>\n",
       "      <th>GOLGA8M</th>\n",
       "      <th>MARF1</th>\n",
       "      <th>ARHGEF11;ETV3L</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNF676</th>\n",
       "      <th>DPCR1</th>\n",
       "      <th>TP53</th>\n",
       "      <th>FAM161B</th>\n",
       "      <th>GNAS</th>\n",
       "      <th>SMAD4</th>\n",
       "      <th>AGFG1</th>\n",
       "      <th>PIK3CA</th>\n",
       "      <th>PRELID3B</th>\n",
       "      <th>ADRA2B;ADRA2B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Case.ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>006-F5</th>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>048-M1-F5</th>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>008-F0</th>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td>MUT;</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          NCOR2 OR11G2  KRAS MUC4;MUC4 DNHD1 TSPYL1 CBWD1 GOLGA8M MARF1  \\\n",
       "Case.ID                                                                   \n",
       "006-F5     MUT;   MUT;  MUT;            MUT;         MUT;    MUT;  MUT;   \n",
       "048-M1-F5  MUT;   MUT;  MUT;      MUT;         MUT;                MUT;   \n",
       "008-F0     MUT;   MUT;  MUT;      MUT;  MUT;   MUT;  MUT;    MUT;  MUT;   \n",
       "\n",
       "          ARHGEF11;ETV3L      ...      ZNF676 DPCR1  TP53 FAM161B  GNAS SMAD4  \\\n",
       "Case.ID                       ...                                               \n",
       "006-F5              MUT;      ...                                        MUT;   \n",
       "048-M1-F5           MUT;      ...                    MUT;          MUT;  MUT;   \n",
       "008-F0              MUT;      ...              MUT;          MUT;  MUT;         \n",
       "\n",
       "          AGFG1 PIK3CA PRELID3B ADRA2B;ADRA2B  \n",
       "Case.ID                                        \n",
       "006-F5            MUT;     MUT;          MUT;  \n",
       "048-M1-F5                                      \n",
       "008-F0     MUT;                                \n",
       "\n",
       "[3 rows x 50 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oncoprint_consensus_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COSMIC Mutational Similarity\n",
    "\n",
    "Output mutational similarity data for all replicates and consensus samples. The COSMIC mutational similarity is built from a (0,1) sample by COSMIC mutation matrix.\n",
    "\n",
    "### For All Replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All COSMIC mutations: 3259\n",
      "Unique COSMIC mutations: 1141\n"
     ]
    }
   ],
   "source": [
    "# How many COSMIC mutation IDs are in the entire set and how many are unique?\n",
    "print('All COSMIC mutations: {}'.format(cosmic_all_df.shape[0]))\n",
    "unique_cosmic_ids = set(cosmic_all_df['cosmic70'])\n",
    "print('Unique COSMIC mutations: {}'.format(len(unique_cosmic_ids)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_id = []\n",
    "cosmic_similarity_list = []\n",
    "for variant_file in os.listdir(variant_file_path):\n",
    "    # Load and subset file to only variants in the COSMIC db\n",
    "    variant_df = pd.read_table(os.path.join(variant_file_path, variant_file), index_col=0)\n",
    "    variant_sub_df = variant_df[variant_df['cosmic70'] != '.']\n",
    "    \n",
    "    # Define membership in COSMIC IDs\n",
    "    cosmic_class = [1 if x in variant_sub_df['cosmic70'].tolist() else 0\n",
    "                    for x in unique_cosmic_ids]\n",
    "    \n",
    "    # Store results\n",
    "    cosmic_similarity_list.append(cosmic_class)\n",
    "    case_id.append(variant_file.replace('_001_processed_variants.tsv.bz2', ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate COSMIC id membership data (for downstream similarity matrix)\n",
    "cosmic_common_file = os.path.join('results', 'cosmic_similarity_replicates.tsv')\n",
    "cosmic_common_df = (\n",
    "    pd.DataFrame(cosmic_similarity_list,\n",
    "                 index=case_id,\n",
    "                 columns=unique_cosmic_ids)\n",
    "    )\n",
    "cosmic_common_df.index.name = 'Case.ID'\n",
    "cosmic_common_df.to_csv(cosmic_common_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Consensus samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_id_consensus = []\n",
    "cosmic_similarity_consensus_list = []\n",
    "for sample_id in set(cosmic_all_df['final_id']):\n",
    "    # Subset file to given sample ID\n",
    "    variant_sub_df = cosmic_all_df.query('final_id == @sample_id')\n",
    "    \n",
    "    # Define membership in COSMIC IDs\n",
    "    cosmic_class = [1 if x in variant_sub_df['cosmic70'].tolist() else 0\n",
    "                    for x in unique_cosmic_ids]\n",
    "    \n",
    "    # Store results\n",
    "    cosmic_similarity_consensus_list.append(cosmic_class)\n",
    "    case_id_consensus.append(sample_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_cosmic_consensus_file = os.path.join('results', 'cosmic_similarity_merged.tsv')\n",
    "cosmic_common_consensus_df = pd.DataFrame(cosmic_similarity_consensus_list,\n",
    "                                          index=case_id_consensus,\n",
    "                                          columns=unique_cosmic_ids)\n",
    "cosmic_common_consensus_df.index.name = 'Case.ID'\n",
    "cosmic_common_consensus_df.to_csv(common_cosmic_consensus_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What about prefiltered variants (i.e. before COSMIC filtering)\n",
    "\n",
    "Observed merged samples with cosmic similarity only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gway/anaconda3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGene</th>\n",
       "      <th>Gene.refGene</th>\n",
       "      <th>GeneDetail.refGene</th>\n",
       "      <th>ExonicFunc.refGene</th>\n",
       "      <th>AAChange.refGene</th>\n",
       "      <th>...</th>\n",
       "      <th>phyloP20way_mammalian</th>\n",
       "      <th>phastCons7way_vertebrate</th>\n",
       "      <th>phastCons20way_mammalian</th>\n",
       "      <th>SiPhy_29way_logOdds</th>\n",
       "      <th>Otherinfo</th>\n",
       "      <th>final_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>het</th>\n",
       "      <th>quality</th>\n",
       "      <th>depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>17379</td>\n",
       "      <td>17379</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>MIR6859-1;MIR6859-2;MIR6859-3;MIR6859-4</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>het\\t.\\t34</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>139058</td>\n",
       "      <td>139058</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>LOC729737</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>het\\t.\\t125</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>664844</td>\n",
       "      <td>664844</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ncRNA_exonic</td>\n",
       "      <td>LOC100133331</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>het\\t.\\t14</td>\n",
       "      <td>008-M2-F0</td>\n",
       "      <td>019-F0</td>\n",
       "      <td>het</td>\n",
       "      <td>.</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr   Start     End Ref Alt  Func.refGene  \\\n",
       "0   1   17379   17379   G   A  ncRNA_exonic   \n",
       "1   1  139058  139058   C   G  ncRNA_exonic   \n",
       "2   1  664844  664844   T   C  ncRNA_exonic   \n",
       "\n",
       "                              Gene.refGene GeneDetail.refGene  \\\n",
       "0  MIR6859-1;MIR6859-2;MIR6859-3;MIR6859-4                  .   \n",
       "1                                LOC729737                  .   \n",
       "2                             LOC100133331                  .   \n",
       "\n",
       "  ExonicFunc.refGene AAChange.refGene  ...  phyloP20way_mammalian  \\\n",
       "0                  .                .  ...                      .   \n",
       "1                  .                .  ...                      .   \n",
       "2                  .                .  ...                      .   \n",
       "\n",
       "  phastCons7way_vertebrate phastCons20way_mammalian SiPhy_29way_logOdds  \\\n",
       "0                        .                        .                   .   \n",
       "1                        .                        .                   .   \n",
       "2                        .                        .                   .   \n",
       "\n",
       "     Otherinfo   final_id sample_name  het quality depth  \n",
       "0   het\\t.\\t34  008-M2-F0      019-F0  het       .    34  \n",
       "1  het\\t.\\t125  008-M2-F0      019-F0  het       .   125  \n",
       "2   het\\t.\\t14  008-M2-F0      019-F0  het       .    14  \n",
       "\n",
       "[3 rows x 60 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load all prefiltered cosmic variants called in this dataset\n",
    "# This file was generated in filter_variants.ipynb\n",
    "file = os.path.join('results', 'all_cosmic_prefiltered_variants.tsv')\n",
    "cosmic_prefiltered_df = pd.read_table(file)\n",
    "cosmic_prefiltered_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NONE;NONE                5484\n",
       "NBPF25P                   963\n",
       "OR4E1;LINC02332           938\n",
       "DNAH5                     858\n",
       "RYR2                      732\n",
       "SYNE2                     719\n",
       "MIR1268A                  696\n",
       "PIKFYVE                   685\n",
       "TG                        651\n",
       "CUBN                      621\n",
       "LRP1B                     618\n",
       "CNTNAP3P2;XLOC_007697     614\n",
       "LOC100996724              600\n",
       "OR4A47;TRIM49B            586\n",
       "CSMD3                     584\n",
       "NONE;LOC654342            581\n",
       "RYR3                      577\n",
       "SYNE1                     571\n",
       "LOC440040;OR4C13          563\n",
       "NONE;CHEK2P2              560\n",
       "Name: Gene.refGene, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paad_prefiltered_genes = cosmic_prefiltered_df['Gene.refGene'].value_counts().head(top_n).index.tolist()\n",
    "cosmic_prefiltered_df['Gene.refGene'].value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only consider mutations that change amino acid sequence\n",
    "cosmic_prefiltered_df = cosmic_prefiltered_df[cosmic_prefiltered_df['AAChange.refGene'] != '.']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All prefiltered COSMIC mutations: 56516\n",
      "Unique nonsilent mutations: 29191\n"
     ]
    }
   ],
   "source": [
    "# How many prefiltered COSMIC mutation IDs are in the entire set and how many are unique?\n",
    "print('All prefiltered COSMIC mutations: {}'.format(cosmic_prefiltered_df.shape[0]))\n",
    "unique_mutations = set(cosmic_prefiltered_df['AAChange.refGene'])\n",
    "print('Unique nonsilent mutations: {}'.format(len(unique_mutations)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "case_id_consensus = []\n",
    "cosmic_similarity_consensus_list = []\n",
    "for sample_id in set(cosmic_prefiltered_df['final_id']):\n",
    "    # Subset file to given sample ID\n",
    "    variant_sub_df = cosmic_prefiltered_df.query('final_id == @sample_id')\n",
    "    \n",
    "    # Define membership in COSMIC IDs\n",
    "    cosmic_class = [1 if x in variant_sub_df['AAChange.refGene'].tolist() else 0\n",
    "                    for x in unique_mutations]\n",
    "    \n",
    "    # Store results\n",
    "    cosmic_similarity_consensus_list.append(cosmic_class)\n",
    "    case_id_consensus.append(sample_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_cosmic_consensus_file = os.path.join('results', 'cosmic_prefiltered_similarity_merged.tsv')\n",
    "cosmic_common_consensus_df = pd.DataFrame(cosmic_similarity_consensus_list,\n",
    "                                          index=case_id_consensus,\n",
    "                                          columns=unique_mutations)\n",
    "cosmic_common_consensus_df.index.name = 'Case.ID'\n",
    "cosmic_common_consensus_df.to_csv(common_cosmic_consensus_file, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
